### DISCUSSION TITLE
Get amount of words in an article

### DISCUSSION THREAD

**Discussion by dbzx10299** [nuxt/content#2795] [Q&A]
I created a time to read function and I was wondering if there was a way to get the amount of words or characters in an article. I am confused because when I look at the queried content it returns to me an AST.

**Comment by oneminch** [comment]
One possible approach is to use [hooks](https://content.nuxt.com/recipes/hooks#contentfilebeforeparse). 

```ts
// server/plugins/content.ts
export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook('content:file:beforeParse', (file) => {
    if (file._id.endsWith('.md')) {
      // counting logic
    }
  })
})
```

**Reply by dbzx10299** [reply]
I did try exactly this. Unified has a unist util size and you pass in the Ast but it doesn't get individual text and needs to use the `afterParse` hook. The reason its a pain with `beforeParse` is from trying to exclude all the mdc components from being counted as words

**Reply by oneminch** [reply]
Have you tried using RegEx in `beforeParse`? Match the markdown specific syntax (`>`, `**`) and MDC component syntax and replace it with an empty string. Then do the counting. I would approach it that way. 

```js
function countWords(text) {
  // Remove markdown syntax using regex
  // Modify regex so it matches MDC syntax as well
  const plainText = text.replace(/[#*_>\-`~]/g, '');
  // Split the text by spaces and filter out empty strings
  const words = plainText.split(/\s+/).filter(word => word.length > 0);
  return words.length;
}

const markdownText = "This is a **markdown** text with *various* elements.";
console.log(countWords(markdownText)); // Output: 7
```

**Reply by dbzx10299** [reply]
Yes I did try something similar but I ran into a different problem that i couldnt set the result as a property on the file object. I was only able to do that with the afterParse hook

**Reply by oneminch** [reply]
I see. In that case, another solution would be to do the counting logic inside the `beforeParse` hook, append the word count directly to the raw body text (since that is possible), and then access it either from your page or as the last node inside the `afterParse` hook.

```ts
// server/plugins/content.ts
function countWords(str: string): number {
  // Modify regex so it matches MDC syntax as well
  const plainText = str.replace(/[#*_>\-`~]/g, "");
  const words = plainText.split(/\s+/).filter((word) => word.length > 0);
  return words.length;
}

export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook("content:file:beforeParse", (file) => {
    if (file._id.endsWith(".md")) {
      // Get word count from raw body
      const wordCount = countWords(file.body);

      // Append word count as text to file body
      // '\n\n' is important for it to be treated as a separate node
      file.body += `\n\n${wordCount}`;
    }
  });

  nitroApp.hooks.hook("content:file:afterParse", (file) => {
    // Get last markdown node which contains word count
    const lastChild = file.body?.children?.at(-1)?.children?.at(0);
    const wordCount = lastChild?.value ?? "";

    // Remove last node from body
    // so it doesn't show up in the page
    file.body?.children.splice(-1, 1);

    // Attach word count to file object
    file.wordCount = wordCount;
  });
});
```

```vue
<ContentDoc>
  <template #default="{ doc }">
    <h1 class="mb-4">
      {{ doc.title }}
    </h1>

    <p>Word Count: {{ doc.wordCount }}</p>

    <!-- Main Content -->
    <ContentRenderer :value="doc" />
  </template>
  <template #not-found>
    <p>No content.</p>
  </template>
</ContentDoc>
```

If this is too much, the only other option seems to be to recursively count the words inside the `file.body.children` object, which isn't ideal.

**Reply by oneminch** [reply]
The raw body text should be accessible from the `afterParse` hook as well for scenarios like this.

**Reply by dbzx10299** [reply]
I wanted to do this with unist util but that creates an AST and the children nodes just contain and entire string of text and it doesnt seem like a clean way to break that up. What do you mean by raw body text?

**Reply by oneminch** [reply]
By raw body text, I meant unparsed Markdown text. I'm creating an issue in this repo requesting the raw body text to be accessible from the `afterParse` hook probably as `file.rawBody`.

Did you see [this approach](https://github.com/nuxt/content/discussions/2795#discussioncomment-10847395)?

**Reply by dbzx10299** [reply]
Yes I did. I created a function to take in text and estimate reading time and I was surprised what a pain it was. I can find ways to do this but the issue I‘m dealing with is it‘s brittle. It would be easy enough to make and it would be easy to break

**Reply by oneminch** [reply]
This module has great potential. Hopefully, such features will get built into the module itself in the future. Until then, we stick to workarounds 😄.
